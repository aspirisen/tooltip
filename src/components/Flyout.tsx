import React from "react";
import { TooltipPlacement } from "../core/Tooltip.types";
import { getPosition } from "../core/getPosition";
import { useClientRect } from "../core/useClientRect";
import cn from "../styles/Flyout.module.css";

export interface FlyoutProps {
  targetRect: DOMRect;
  placement: TooltipPlacement;
}

export const Flyout: React.FC<FlyoutProps> = props => {
  const [tooltipRect, ref] = useClientRect<HTMLDivElement>([props.targetRect]);

  const styles: React.CSSProperties =
    props.targetRect && tooltipRect
      ? getPosition(props.placement, props.targetRect, tooltipRect)
      : { visibility: "hidden" };

  return (
    <div ref={ref} className={cn.flyout} style={styles}>
      <div className={cn.body}>{props.children}</div>
    </div>
  );
};
