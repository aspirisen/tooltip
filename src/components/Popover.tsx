import React from "react";
import { TooltipPlacement } from "../core/Tooltip.types";
import { Flyout } from "./Flyout";

export interface PopoverProps {
  target: React.RefObject<HTMLElement>;
  placement: TooltipPlacement;
}

export const Popover: React.FC<PopoverProps> = props => {
  const [isOver, setIsOver] = React.useState(false);

  React.useEffect(() => {
    const enter = () => setIsOver(true);
    const leave = () => setIsOver(false);

    props.target.current?.addEventListener("mouseenter", enter);
    props.target.current?.addEventListener("mouseleave", leave);

    return () => {
      props.target.current?.removeEventListener("mouseenter", enter);
      props.target.current?.removeEventListener("mouseleave", leave);
    };
  }, [props.target.current]);

  const targetRect = props.target.current?.getBoundingClientRect();

  if (!isOver || !targetRect) {
    return null;
  }

  return (
    <Flyout placement={props.placement} targetRect={targetRect}>
      {props.children}
    </Flyout>
  );
};
