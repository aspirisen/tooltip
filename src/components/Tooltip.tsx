import React from "react";
import ReactDOM from "react-dom";
import { Popover, PopoverProps } from "./Popover";

export interface TooltipProps<T extends HTMLElement> extends Pick<PopoverProps, "placement"> {
  content: React.ReactNode;
  children: (ref: React.RefObject<T>) => React.ReactNode;
}

export function Tooltip<T extends HTMLElement>(props: TooltipProps<T>) {
  const ref = React.createRef<T>();
  const targetElement = props.children(ref);

  return (
    <>
      {targetElement}
      {ReactDOM.createPortal(
        <Popover placement={props.placement} target={ref}>
          {props.content}
        </Popover>,
        document.body
      )}
    </>
  );
}
