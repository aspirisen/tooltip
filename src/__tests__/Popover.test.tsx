import React from "react";
import { render, act } from "@testing-library/react";
import { Popover } from "../components/Popover";

describe("Popover", () => {
  it("should render Popover", async () => {
    let enter: Function;

    const element = {
      getBoundingClientRect: () => ({
        left: 0,
        top: 0,
        bottom: 40,
        height: 40,
        width: 40,
        x: 0,
        y: 0
      }),
      addEventListener: (type: string, callback: EventListener) => {
        if (type === "mouseenter") {
          enter = callback;
        }
      }
    } as HTMLDivElement;

    element.removeEventListener = jest.fn();

    const target = { current: element };

    const wrapper = render(
      <Popover placement="bottom" target={target}>
        <div>Some Content</div>
      </Popover>
    );

    act(() => {
      enter();
    });

    expect(wrapper.container).toMatchSnapshot();
  });

  it("should open Popover on mouse enter", async () => {
    let enter: Function;

    const element = {
      getBoundingClientRect: () => ({
        left: 0,
        top: 0,
        bottom: 40,
        height: 40,
        width: 40,
        x: 0,
        y: 0
      }),
      addEventListener: (type: string, callback: EventListener) => {
        if (type === "mouseenter") {
          enter = callback;
        }
      }
    } as HTMLDivElement;

    element.removeEventListener = jest.fn();

    const target = { current: element };

    const id = "content";

    const wrapper = render(
      <Popover placement="bottom" target={target}>
        <div data-testid={id}>Some Content</div>
      </Popover>
    );

    act(() => {
      enter();
    });

    expect(wrapper.getByTestId(id)).toBeDefined();
  });

  it("should close tooltip on mouse leave", async () => {
    let enter: Function;
    let leave: Function;

    const element = {
      getBoundingClientRect: () => ({
        left: 0,
        top: 0,
        bottom: 40,
        height: 40,
        width: 40,
        x: 0,
        y: 0
      }),
      addEventListener: (type: string, callback: EventListener) => {
        if (type === "mouseenter") {
          enter = callback;
        }

        if (type === "mouseleave") {
          leave = callback;
        }
      }
    } as HTMLDivElement;

    element.removeEventListener = jest.fn();

    const target = { current: element };

    const id = "content";

    const wrapper = render(
      <Popover placement="bottom" target={target}>
        <div data-testid={id}>Some Content</div>
      </Popover>
    );

    act(() => {
      enter();
    });

    expect(wrapper.getByTestId(id)).toBeDefined();

    act(() => {
      leave();
    });

    expect(() => wrapper.getByTestId(id)).toThrow();
  });
});
