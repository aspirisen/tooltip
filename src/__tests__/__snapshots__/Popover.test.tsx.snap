// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover should render Popover 1`] = `
<div>
  <div
    class="flyout"
    style="top: 40px; left: 20px;"
  >
    <div
      class="body"
    >
      <div>
        Some Content
      </div>
    </div>
  </div>
</div>
`;
